pkg/policy/mapstate.go

```diff
@@ -114,55 +114,91 @@ func (e *MapStateEntry) IsRedirectEntry() bool {
 	return e.ProxyPort != 0
 }
 
-// Equal returns true of two entries are equal
-func (e *MapStateEntry) Equal(o *MapStateEntry) bool {
+// DatapathEqual returns true of two entries are equal in the datapath's PoV,
+// i.e., both Deny and ProxyPort are the same for both entries.
+func (e *MapStateEntry) DatapathEqual(o *MapStateEntry) bool {
 	if e == nil || o == nil {
 		return e == o
 	}
 
-	return e.IsDeny == o.IsDeny && e.ProxyPort == o.ProxyPort && e.DerivedFromRules.Equals(o.DerivedFromRules)
+	return e.IsDeny == o.IsDeny && e.ProxyPort == o.ProxyPort
 }
 
 // DenyPreferredInsert inserts a key and entry into the map by given preference
 // to deny entries, and L3-only deny entries over L3-L4 allows.
-func (keys MapState) DenyPreferredInsert(key Key, entry MapStateEntry) {
-	// If it is a deny policy, it will take precedence over any other key.
-	if entry.IsDeny {
-		// If there is an allow or a deny L3-L4 and the new deny is a L3-Only,
-		// the new deny takes precedence over all L3-L4 for each the L3 is the
-		// same on both the existing rules and new deny rule.
-		if key.DestPort == 0 {
+func (keys MapState) DenyPreferredInsert(newKey Key, newEntry MapStateEntry) {
+	if newEntry.IsDeny {
+		// case for an existing allow L4-only and we are inserting deny L3-only
+		if newKey.DestPort == 0 && newKey.Nexthdr == 0 && newKey.Identity != 0 {
+			l4OnlyAllows := MapState{}
+			for k, v := range keys {
+				if newKey.TrafficDirection == k.TrafficDirection &&
+					!v.IsDeny &&
+					k.Identity == 0 {
+					// create a deny L3-L4 with the same allowed L4 port and proto
+					newKeyCpy := newKey
+					newKeyCpy.DestPort = k.DestPort
+					newKeyCpy.Nexthdr = k.Nexthdr
+					keys[newKeyCpy] = newEntry
+
+					l4OnlyAllows[k] = v
+				}
+			}
+			// Delete all L3-L4 if we are inserting a deny L3-only and
+			// there aren't allow L4-only for the existing deny L3-L4
 			for k := range keys {
-				if k.TrafficDirection == key.TrafficDirection &&
-					k.Identity == key.Identity {
-					delete(keys, k)
+				if k.TrafficDirection == newKey.TrafficDirection &&
+					k.DestPort != 0 && k.Nexthdr != 0 &&
+					k.Identity == newKey.Identity {
+
+					kCpy := k
+					kCpy.Identity = 0
+					if _, ok := l4OnlyAllows[kCpy]; !ok {
+						delete(keys, k)
+					}
 				}
 			}
+		} else {
+			// If there is a Deny that is an L3-only, all L3-L4 denies won't be
+			// accepted because a L3-Deny takes precedence.
```

Comment:
`won't be accepted` is a bit vague, and describing "L3 deny takes precedence" is a bit confusing since at the datapath level, the evaluation order is fixed to L3+L4 -> L4 -> L3 -> all, and this function is handling map state which will be directly checked by that datapath evaluation logic.

I think maybe it's sufficient to just use imperative language?

```suggestion
			// Do not insert 'newKey' if the map state already denies traffic
			// which is a superset of (or equal to) 'newKey'
```

Reading it this way, I think it's clearer to understand that you don't want to insert an entry that would allow traffic if the mapstate currently determines that that traffic must be denied.